tachyon/zk/lookup/logup/prover.h

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_LOOKUP_LOGUP_PROVER_H_
+#define TACHYON_ZK_LOOKUP_LOGUP_PROVER_H_
+
+#include <vector>
+
+#include "tachyon/crypto/commitments/polynomial_openings.h"
+#include "tachyon/zk/base/blinded_polynomial.h"
+#include "tachyon/zk/base/entities/prover_base.h"
+#include "tachyon/zk/lookup/halo2/opening_point_set.h"
+#include "tachyon/zk/lookup/lookup_argument.h"
+#include "tachyon/zk/lookup/proving_evaluator.h"
+#include "tachyon/zk/plonk/base/multi_phase_ref_table.h"
+
+namespace tachyon::zk::lookup::logup {
+
+template <typename Poly, typename Evals>
+class Prover {
+ public:
+  using F = typename Poly::Field;
+
+  const std::vector<std::vector<Evals>>& compressed_inputs_vec() const {
+    return compressed_inputs_vec_;
+  }
+
+  const std::vector<Evals>& compressed_tables() const {
+    return compressed_tables_;
+  }
+
+  const std::vector<BlindedPolynomial<Poly, Evals>>& m_polys() const {
+    return m_polys_;
+  }
+
+  const std::vector<BlindedPolynomial<Poly, Evals>>& grand_sum_polys() const {
+    return grand_sum_polys_;
+  }
+
+  template <typename Domain>
+  static void BatchCompressPairs(
+      std::vector<Prover>& lookup_provers, const Domain* domain,
+      const std::vector<Argument<F>>& arguments, const F& theta,
+      const std::vector<plonk::MultiPhaseRefTable<Evals>>& tables);
+
+  template <typename PCS>
+  static void BatchComputeMPolys(std::vector<Prover>& lookup_provers,
+                                 ProverBase<PCS>* prover) {
+    for (Prover& lookup_prover : lookup_provers) {
+      lookup_prover.ComputeMPolys(prover);
+    }
+  }
+
+  constexpr static size_t GetNumMPolysCommitments(
+      const std::vector<Prover>& lookup_provers) {
+    if (lookup_provers.empty()) return 0;
+    return lookup_provers.size() * lookup_provers[0].m_polys_.size();
+  }
+
+  template <typename PCS>
+  static void BatchCommitMPolys(const std::vector<Prover>& lookup_provers,
+                                ProverBase<PCS>* prover, size_t& commit_idx);
```

Comment:
```suggestion
                                ProverBase<PCS>* prover, size_t commit_idx);
```
Shouldn't we pass by value?